geode-core/src/main/java/org/apache/geode/security/AuthenticationExpiredException.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.security;
+
+/**
+ * This exception is thrown by the SecurityManager's authorize method to indicate that the
+ * authentication has expired.
+ */
```

Comment:
Let's take this small surface area to fix a wrong. Not all exceptions should be unchecked. It is bad pattern to have unchecked exceptions thrown all over the place. It makes it hard for authors to know when and where to catch exceptions. 

`AuthenticationExpiredException` should be a checked exception. To do so we need to correct the wrong documented in the `GemFireException` where it was derived from `RuntimeException`. We should define a new base exception `GeodeException` that derives from the checked `Exception`. This new exception should derive from `GeodeException` and be checked.

Reply:
wouldn't this break the API?

Reply:
Nope. Exceptions on an interface are optional on the implementation. Checked exceptions are no exception. Current implementations with both compile and run without issue. This particular interface is an SPI which means the only callers of this API are internal to Geode, meaning only Geode code would be subject to the checked exception being handled.

Reply:
@kohlmu-pivotal seems to indicate that it does in the RFC. Will let him comment on this.

Reply:
this is from Oracle java doc site:
13.4.21. Method and Constructor Throws
Changes to the throws clause of methods or constructors do not break compatibility with pre-existing binaries; these clauses are checked only at compile time.

Reply:
> this is from Oracle java doc site:
> 13.4.21. Method and Constructor Throws
> Changes to the throws clause of methods or constructors do not break compatibility with pre-existing binaries; these clauses are checked only at compile time.

Yes, only the calling code at compile time would be exposed to this change and since the only caller is the internals we can safely make this SPI change.

Reply:
There is a `GemfireCheckedExcpetion`, Do we want to do AuthenticationExpiredException -> GemfireCheckedException, or add a `GemfireCheckedSecurityException` in between?

Reply:
Actually, none of the callers needs to handle this exception, it just gets thrown all the way to the client, and when client got it, it gets wrapped in the `ServerOperationException` as a cause, wondering what's really the benefit of making it a checked exception.....

Reply:
We better be catching it to remove the long ID from the map on the server.

Reply:
Oh, yeah, that's a story we haven't touched upon yet

Reply:
> There is a `GemfireCheckedExcpetion`, Do we want to do AuthenticationExpiredException

I don't see the point in continuing to use the GemFire named symbols. It's about time we start making migration away where we can.

Reply:
> Oh, yeah, that's a story we haven't touched upon yet

And that is why making it checked is important. It forces us to actually deal with the exception and not forget it could be thrown.

Reply:
@jake Turns out the `authorize` has to throw an unchecked exception or internally we have to turn it into an unchecked exception anyway, since this method is called by Shiro's `AuthrizingRealm` (see our implementation `CustomAuthRealm`), it's interface `isPermitted` doesn't throw any exception, only returns a boolean

Reply:
Wow, another point against Shiro. Ok.

Reply:
At the very least then I don't think it makes sense for this exception to extend `AuthenticationRequiredException`. Is there a relationship there I am not thinking of?

Reply:
Not from a coding perspective, I just though it's a natural extension, "auth expired means auth required". I can change it  to extend `GemfireSecurityException"